services/QuillLMS/app/services/diagnostic_insights/skill_practice_by_classroom.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module DiagnosticInsights
+  class SkillPracticeByClassroom < ApplicationService
+    class UnknownDiagnosticType < StandardError; end
+
+    attr_reader :classroom_id, :diagnostic_type, :start_date, :end_date
+
+    def initialize(classroom_id, diagnostic_type, start_date: nil, end_date: nil)
+      @classroom_id = classroom_id
+      @diagnostic_type = diagnostic_type
+      @start_date = start_date || DEFAULT_START_DATE
+      @end_date = end_date || Float::INFINITY
+    end
+
+    def run
+      raise UnknownDiagnosticType, "Unknown diagnostic_type: #{diagnostic_type}" if unknown_diagnostic_type?
+
+      students.map do |student|
+        {
+          user_id: student.id,
+          user_name: student.name,
+          skills: format_skills_for_student(student)
+        }
+      end
+    end
+
+    private def format_skills_for_student(student)
+      student_activity_sessions = activity_sessions_by_student[student.id]
+
+      return [] unless student_activity_sessions
+
+      format_skills_for_activity_sessions(student_activity_sessions)
+    end
+
+    private def format_skills_for_activity_sessions(student_activity_sessions)
+      student_activity_sessions
+        .filter { |activity_session| activity_session.skill_performance&.first&.fetch('skill_name', nil).present? }
+        .group_by { |activity_session| activity_session.skill_performance.first['skill_name'] }
+        .map do |skill_name, activity_sessions|
+          generate_practice_payload(skill_name, activity_sessions)
+        end
+    end
+
+    private def generate_practice_payload(skill_name, activity_sessions)
+      {
+        skill_name:,
+        activity_count: activity_sessions.length,
+        timespent: activity_sessions.map(&:timespent).sum,
+        last_practiced: activity_sessions.map(&:completed_at).max,
+        percentages: activity_sessions.map(&:percentage)
+      }
+    end
+
+    private def students
+      @students ||= User
+        .joins(students_classrooms: :classroom_unscoped)
+        .where(students_classrooms: { classrooms: { id: classroom_id } })
```

Comment:
You can save a join here.
```suggestion
        .joins(:students_classrooms)
        .where(students_classrooms: { classroom_id: } )
```